src/Datagrid/Datagrid.php

```diff
@@ -330,6 +291,64 @@ public function getPaginationParameters(int $page): array
         return ['filter' => $values];
     }
 
+    private function applyFilters(array $data): void
+    {
+        foreach ($this->getFilters() as $name => $filter) {
+            $this->values[$name] = isset($this->values[$name]) ? $this->values[$name] : null;
+            $filterFormName = $filter->getFormName();
+            if (isset($this->values[$filterFormName]['type'], $this->values[$filterFormName]['value']) &&
+                ('' !== $this->values[$filterFormName]['type'] || '' !== $this->values[$filterFormName]['value'])
+            ) {
+                $filter->apply($this->query, $data[$filterFormName]);
+            }
+        }
+    }
+
+    private function applySorting(): void
+    {
+        if (isset($this->values['_sort_by'])) {
+            if (!$this->values['_sort_by'] instanceof FieldDescriptionInterface) {
+                throw new UnexpectedTypeException($this->values['_sort_by'], FieldDescriptionInterface::class);
+            }
+
+            if ($this->values['_sort_by']->isSortable()) {
+                $this->query->setSortBy(
+                    $this->values['_sort_by']->getSortParentAssociationMapping(),
+                    $this->values['_sort_by']->getSortFieldMapping()
+                );
+
+                $this->values['_sort_order'] = $this->values['_sort_order'] ?? 'ASC';
+                $this->query->setSortOrder($this->values['_sort_order']);
+            }
+        }
+    }
+
+    private function getMaxPerPage(int $default): int
+    {
+        if (isset($this->values['_per_page'])) {
+            if (isset($this->values['_per_page']['value'])) {
+                return (int) $this->values['_per_page']['value'];
```

Comment:
I assume that this data comes from a request and it contains a string with a number, not a integer. Therefore, i am doing explicit type casting.

Reply:
It's never too safe